consensus/expected.go

```diff
@@ -100,50 +101,154 @@ type Expected struct {
+}
+
+type BlockStateValidator interface {
+	ValidateState(ctx context.Context, st state.Tree, cur, parent types.TipSet) error
+}
+
+type BlockValidatorClock interface {
+	Now() types.Uint64
+	BlockTime() types.Uint64
+}
+
+type ExpectedBlockValidatorClock struct {
+	blockTime time.Duration
+}
+
+func NewExpectedBlockValidatorClock(blockTime time.Duration) BlockValidatorClock {
+	return &ExpectedBlockValidatorClock{
+		blockTime: blockTime,
 	}
 }
 
-// NewValidTipSet creates a new tipset from the input blocks that is guaranteed
-// to be valid. It operates by validating each block and further checking that
-// this tipset contains only blocks with the same heights, parent weights,
-// and parent sets.
-func (c *Expected) NewValidTipSet(ctx context.Context, blks []*types.Block) (types.TipSet, error) {
-	for _, blk := range blks {
-		if err := c.validateBlockStructure(ctx, blk); err != nil {
-			return nil, err
-		}
+func (ebvc *ExpectedBlockValidatorClock) Now() types.Uint64 {
+	return types.Uint64(time.Now().Unix())
+}
+
+func (ebvc *ExpectedBlockValidatorClock) BlockTime() types.Uint64 {
+	return types.Uint64(ebvc.blockTime.Seconds())
+}
+
+type ExpectedBlockValidator struct {
+	BlockValidatorClock
+}
+
+func NewExpectedBlockValidator(clock BlockValidatorClock) BlockValidator {
+	return &ExpectedBlockValidator{
+		BlockValidatorClock: clock,
 	}
-	return types.NewTipSet(blks...)
 }
 
-// ValidateBlockStructure verifies that this block, on its own, is structurally and
+// ValidateSyntax verifies that this block, on its own, is structurally and
 // cryptographically valid. This means checking that all of its fields are
 // properly filled out and its signatures are correct. Checking the validity of
 // state changes must be done separately and only once the state of the
 // previous block has been validated. TODO: not yet signature checking
-func (c *Expected) validateBlockStructure(ctx context.Context, b *types.Block) error {
+func (ebv *ExpectedBlockValidator) ValidateSyntax(ctx context.Context, ts types.TipSet) error {
 	// TODO: validate signature on block
-	if !b.StateRoot.Defined() {
-		return fmt.Errorf("block has nil StateRoot")
+	// TODO: valudate fileds
+	blks := ts.ToSlice()
+	one := blks[0]
+	for _, b := range blks {
+		if !b.StateRoot.Defined() {
+			return fmt.Errorf("block has nil StateRoot")
+		}
+
+		if b.Timestamp > ebv.Now() {
+			return errors.Errorf("block: %s was generated by miner: %s too far in the future", b.Cid(), b.Miner.String())
+		}
+		// validate same parents
+		if b.Parents.String() != one.Parents.String() {
+			return errors.Errorf("invalid parents for block: %s parents: %d & block: %s, parents: %d",
+				b.Cid().String(), b.Parents.String(), one.Cid().String(), one.Parents.String())
+		}
+
+		// validate same height
+		if b.Height != one.Height {
+			return errors.Errorf("invalid height for block: %s height: %d & block: %s, height: %d",
+				b.Cid().String(), b.Height, one.Cid().String(), one.Height)
+		}
```

Comment:
When https://github.com/filecoin-project/go-filecoin/pull/2837 lands, the tipset consistency checks are redundant (I already copied from the same place you took these). I do think this should be limited to single block validation so it is usable before a block has been joined with others into a tipset.